src/platform_impl/linux/wayland/window/mod.rs

```diff
@@ -86,33 +91,64 @@ impl Window {
         attributes: WindowAttributes,
         platform_attributes: PlatformAttributes,
     ) -> Result<Self, RootOsError> {
-        let surface = event_loop_window_target
+        let surface;
+        let mut fractional_scale = None;
+        let mut viewport = None;
+        let mut scale_factor = 1.0;
+
+        let viewporter = event_loop_window_target.env.get_global::<WpViewporter>();
+        let fractional_scale_manager = event_loop_window_target
             .env
-            .create_surface_with_scale_callback(move |scale, surface, mut dispatch_data| {
+            .get_global::<WpFractionalScaleManagerV1>();
+
+        if let (Some(vp), Some(fsm)) = (viewporter, fractional_scale_manager) {
+            surface = event_loop_window_target.env.create_surface().detach();
+            let window_id = super::make_wid(&surface);
+            let fs = fsm.get_fractional_scale(&surface);
+            fs.quick_assign(move |_, event, mut dispatch_data| {
+                let wp_fractional_scale_v1::Event::PreferredScale { scale } = event;
                 let winit_state = dispatch_data.get::<WinitState>().unwrap();
-
-                // Get the window that received the event.
-                let window_id = super::make_wid(&surface);
-                let mut window_compositor_update = winit_state
+                winit_state
                     .window_compositor_updates
                     .get_mut(&window_id)
-                    .unwrap();
-
-                // Mark that we need a frame refresh on the DPI change.
+                    .unwrap()
+                    .scale_factor = Some(scale as f64 / 120.0);
                 winit_state
                     .window_user_requests
                     .get_mut(&window_id)
                     .unwrap()
                     .refresh_frame = true;
+            });
+            fractional_scale = Some(fs.detach());
+            viewport = Some(vp.get_viewport(&surface).detach());
```

Comment:
Maybe we should store them in some common struct like

```rust
struct FractionalScalingState {
  viewporter: Viewporter,
  fractional_scaling: FractionalScaling,
}
```

That should simplify a bunch of `Option` handling, since you have either both `None` or both `Some` at the same time.